File: go/space/space.go
// SPDX-License-Identifier: MIT

// Package space implements aether-space: geo-pinned community noticeboards
// (Phase-2 extension). Nodes drop breadcrumbs at geohash coordinates; passing
// devices auto-pull and re-host them for other passersby — fully offline.
//
// Port of the C# reference (AetherNet.Space). Wire format: JSON, transmitted as
// PacketType.SpaceBreadcrumb (40).
package space

import (
	"context"
	"strings"
	"sync"
	"time"
)

// BreadcrumbType is the category of a geo-pinned breadcrumb.
type BreadcrumbType uint8

const (
	// BreadcrumbTypeNotice is a general community notice (default).
	BreadcrumbTypeNotice BreadcrumbType = 0
	// BreadcrumbTypeEmergency bypasses the flood-guard; TTL extended to 720 h.
	BreadcrumbTypeEmergency BreadcrumbType = 1
	// BreadcrumbTypeCommerce is a commercial listing or market offer.
	BreadcrumbTypeCommerce BreadcrumbType = 2
	// BreadcrumbTypeEvent is a local event announcement.
	BreadcrumbTypeEvent BreadcrumbType = 3
	// BreadcrumbTypeJobPosting is a job posting or opportunity.
	BreadcrumbTypeJobPosting BreadcrumbType = 4
)

// EmergencyTtlHours is the fixed TTL applied to Emergency breadcrumbs.
const EmergencyTtlHours = 720

// MinTtlHours / MaxTtlHours bound a non-emergency breadcrumb's lifetime.
const (
	MinTtlHours = 1
	MaxTtlHours = 168
)

// Breadcrumb is a geo-pinned digital notice dropped by a user at a physical
// location. Content is addressed by hash; the breadcrumb carries only metadata.
type Breadcrumb struct {
	ContentHash string // content-service hash of the actual payload
	GeoHash     string // 6-character geohash of the drop location (~1.2 km² cell)
	AnchorUhid  string // UHID of the node that dropped the breadcrumb
	CreatedAt   time.Time
	TtlHours    int
	Type        BreadcrumbType
	Signature   []byte // Ed25519 over (ContentHash + GeoHash + CreatedAt ISO-8601); empty if unsigned
}

// ExpiresAt is CreatedAt + TtlHours.
func (b *Breadcrumb) ExpiresAt() time.Time {
	return b.CreatedAt.Add(time.Duration(b.TtlHours) * time.Hour)
}

// IsExpired reports whether the breadcrumb's TTL has passed.
func (b *Breadcrumb) IsExpired() bool {
	return !time.Now().UTC().Before(b.ExpiresAt())
}

// Service is the aether-space breadcrumb store.
type Service interface {
	// Drop creates a new breadcrumb at geoHash. ttlHours is clamped to
	// [1,168]; Emergency breadcrumbs are fixed at 720 h.
	Drop(ctx context.Context, geoHash, contentHash, anchorUhid string, typ BreadcrumbType, ttlHours int) (*Breadcrumb, error)
	// Scan returns active (non-expired) breadcrumbs near centerGeoHash.
	Scan(ctx context.Context, centerGeoHash string, radiusCells int) ([]Breadcrumb, error)
	// Pin caches and re-hosts a breadcrumb received from a peer.
	Pin(ctx context.Context, breadcrumb *Breadcrumb) error
	// Delete removes a breadcrumb; succeeds only if requestorUhid is the
	// breadcrumb's AnchorUhid (creator-only delete).
	Delete(ctx context.Context, breadcrumb *Breadcrumb, requestorUhid string) (bool, error)
	// PruneExpired drops every expired breadcrumb and returns the count removed.
	PruneExpired() int
}

// InMemoryService is an in-memory Service for testing and single-node use; all
// state is lost on restart. Proximity matching uses a geohash-prefix heuristic.
//
// Not safe to mutate the callbacks concurrently with calls; set them before use.
type InMemoryService struct {
	mu    sync.Mutex
	store map[string]*Breadcrumb // key = ContentHash

	// OnBreadcrumbReceived fires when a breadcrumb is dropped locally or pinned
	// from the mesh. OnBreadcrumbExpired fires when a cached breadcrumb is pruned.
	OnBreadcrumbReceived func(*Breadcrumb)
	OnBreadcrumbExpired  func(*Breadcrumb)
}

// NewInMemoryService constructs an empty in-memory space service.
func NewInMemoryService() *InMemoryService {
	return &InMemoryService{store: make(map[string]*Breadcrumb)}
}

func clampInt(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// Drop implements Service.
func (s *InMemoryService) Drop(ctx context.Context, geoHash, contentHash, anchorUhid string, typ BreadcrumbType, ttlHours int) (*Breadcrumb, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	effectiveTtl := clampInt(ttlHours, MinTtlHours, MaxTtlHours)
	if typ == BreadcrumbTypeEmergency {
		effectiveTtl = EmergencyTtlHours
	}
	crumb := &Breadcrumb{
		ContentHash: contentHash,
		GeoHash:     geoHash,
		AnchorUhid:  anchorUhid,
		CreatedAt:   time.Now().UTC(),
		TtlHours:    effectiveTtl,
		Type:        typ,
	}
	s.mu.Lock()
	s.store[contentHash] = crumb
	s.mu.Unlock()
	if s.OnBreadcrumbReceived != nil {
		s.OnBreadcrumbReceived(crumb)
	}
	return crumb, nil
}

// Scan implements Service.
func (s *InMemoryService) Scan(ctx context.Context, centerGeoHash string, radiusCells int) ([]Breadcrumb, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	// Prefix-based proximity: match the first (6 - radiusCells) chars.
	prefixLen := clampInt(6-radiusCells, 1, 6)
	prefix := centerGeoHash
	if len(centerGeoHash) >= prefixLen {
		prefix = centerGeoHash[:prefixLen]
	}
	lowerPrefix := strings.ToLower(prefix)

	s.mu.Lock()
	defer s.mu.Unlock()
	results := make([]Breadcrumb, 0)
	for _, c := range s.store {
		if !c.IsExpired() && strings.HasPrefix(strings.ToLower(c.GeoHash), lowerPrefix) {
			results = append(results, *c)
		}
	}
	return results, nil
}

// Pin implements Service.
func (s *InMemoryService) Pin(ctx context.Context, breadcrumb *Breadcrumb) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	s.store[breadcrumb.ContentHash] = breadcrumb
	s.mu.Unlock()
	if s.OnBreadcrumbReceived != nil {
		s.OnBreadcrumbReceived(breadcrumb)
	}
	return nil
}

// Delete implements Service.
func (s *InMemoryService) Delete(ctx context.Context, breadcrumb *Breadcrumb, requestorUhid string) (bool, error) {
	if err := ctx.Err(); err != nil {
		return false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	stored, ok := s.store[breadcrumb.ContentHash]
	if !ok {
		return false, nil
	}
	if stored.AnchorUhid != requestorUhid {
		return false, nil // creator-only delete
	}
	delete(s.store, breadcrumb.ContentHash)
	return true, nil
}

// PruneExpired implements Service.
func (s *InMemoryService) PruneExpired() int {
	s.mu.Lock()
	expired := make([]*Breadcrumb, 0)
	for _, c := range s.store {
		if c.IsExpired() {
			expired = append(expired, c)
		}
	}
	for _, c := range expired {
		delete(s.store, c.ContentHash)
	}
	s.mu.Unlock()

	for _, c := range expired {
		if s.OnBreadcrumbExpired != nil {
			s.OnBreadcrumbExpired(c)
		}
	}
	return len(expired)
}

File: rust/src/space.rs
// SPDX-License-Identifier: MIT
//! aether-space: geo-pinned community noticeboards (Phase-2 extension).
//!
//! Nodes drop breadcrumbs at geohash coordinates; passing devices auto-pull and
//! re-host them for other passersby — fully offline. Port of the C# reference
//! (AetherNet.Space). Wire format: JSON, transmitted as PacketType::SpaceBreadcrumb (40).

use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

/// Category of a geo-pinned breadcrumb.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BreadcrumbType {
    Notice = 0,
    Emergency = 1,
    Commerce = 2,
    Event = 3,
    JobPosting = 4,
}

/// Fixed TTL applied to Emergency breadcrumbs.
pub const EMERGENCY_TTL_HOURS: i32 = 720;
/// Bounds for a non-emergency breadcrumb's lifetime (hours).
pub const MIN_TTL_HOURS: i32 = 1;
pub const MAX_TTL_HOURS: i32 = 168;

fn now_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

/// A geo-pinned digital notice dropped at a physical location. Content is
/// addressed by hash; the breadcrumb carries only metadata.
#[derive(Clone, Debug)]
pub struct SpaceBreadcrumb {
    pub content_hash: String,
    pub geo_hash: String,
    pub anchor_uhid: String,
    /// Unix-epoch seconds of creation.
    pub created_at_secs: u64,
    pub ttl_hours: i32,
    pub crumb_type: BreadcrumbType,
    /// Ed25519 signature over (content_hash + geo_hash + created_at); empty if unsigned.
    pub signature: Vec<u8>,
}

impl SpaceBreadcrumb {
    /// Unix-epoch seconds of expiry = created_at_secs + ttl_hours.
    pub fn expires_at_secs(&self) -> u64 {
        self.created_at_secs + (self.ttl_hours.max(0) as u64) * 3600
    }

    /// True once the breadcrumb's TTL has passed.
    pub fn is_expired(&self) -> bool {
        now_secs() >= self.expires_at_secs()
    }
}

/// Optional callback fired on breadcrumb received / expired.
type Callback = Box<dyn Fn(&SpaceBreadcrumb)>;

/// In-memory aether-space breadcrumb store for testing / single-node use; state
/// is lost on drop. Proximity matching uses a geohash-prefix heuristic.
#[derive(Default)]
pub struct InMemorySpaceService {
    store: HashMap<String, SpaceBreadcrumb>, // key = content_hash
    pub on_breadcrumb_received: Option<Callback>,
    pub on_breadcrumb_expired: Option<Callback>,
}

impl InMemorySpaceService {
    pub fn new() -> Self {
        Self::default()
    }

    /// Create a new breadcrumb at `geo_hash`. `ttl_hours` is clamped to [1,168];
    /// Emergency breadcrumbs are fixed at 720 h.
    pub fn drop_crumb(
        &mut self,
        geo_hash: &str,
        content_hash: &str,
        anchor_uhid: &str,
        crumb_type: BreadcrumbType,
        ttl_hours: i32,
    ) -> SpaceBreadcrumb {
        let effective_ttl = if crumb_type == BreadcrumbType::Emergency {
            EMERGENCY_TTL_HOURS
        } else {
            ttl_hours.clamp(MIN_TTL_HOURS, MAX_TTL_HOURS)
        };
        let crumb = SpaceBreadcrumb {
            content_hash: content_hash.to_string(),
            geo_hash: geo_hash.to_string(),
            anchor_uhid: anchor_uhid.to_string(),
            created_at_secs: now_secs(),
            ttl_hours: effective_ttl,
            crumb_type,
            signature: Vec::new(),
        };
        self.store.insert(content_hash.to_string(), crumb.clone());
        if let Some(cb) = &self.on_breadcrumb_received {
            cb(&crumb);
        }
        crumb
    }

    /// Return active (non-expired) breadcrumbs near `center_geo_hash`.
    pub fn scan(&self, center_geo_hash: &str, radius_cells: i32) -> Vec<SpaceBreadcrumb> {
        // Prefix-based proximity: match the first (6 - radius_cells) chars.
        let prefix_len = (6 - radius_cells).clamp(1, 6) as usize;
        let prefix = if center_geo_hash.len() >= prefix_len {
            &center_geo_hash[..prefix_len]
        } else {
            center_geo_hash
        }
        .to_lowercase();
        self.store
            .values()
            .filter(|c| !c.is_expired() && c.geo_hash.to_lowercase().starts_with(&prefix))
            .cloned()
            .collect()
    }

    /// Cache and re-host a breadcrumb received from a peer.
    pub fn pin(&mut self, breadcrumb: SpaceBreadcrumb) {
        let key = breadcrumb.content_hash.clone();
        self.store.insert(key.clone(), breadcrumb);
        if let Some(cb) = &self.on_breadcrumb_received {
            cb(&self.store[&key]);
        }
    }

    /// Creator-only delete: succeeds only if `requestor_uhid` is the breadcrumb's anchor_uhid.
    pub fn delete(&mut self, breadcrumb: &SpaceBreadcrumb, requestor_uhid: &str) -> bool {
        match self.store.get(&breadcrumb.content_hash) {
            Some(stored) if stored.anchor_uhid == requestor_uhid => {
                self.store.remove(&breadcrumb.content_hash);
                true
            }
            _ => false,
        }
    }

    /// Drop every expired breadcrumb; returns the count removed.
    pub fn prune_expired(&mut self) -> i32 {
        let expired: Vec<SpaceBreadcrumb> =
            self.store.values().filter(|c| c.is_expired()).cloned().collect();
        for crumb in &expired {
            self.store.remove(&crumb.content_hash);
            if let Some(cb) = &self.on_breadcrumb_expired {
                cb(crumb);
            }
        }
        expired.len() as i32
    }
}
